src/cowboy_decompress_h.erl

```diff
@@ -0,0 +1,217 @@
+				[] ->
+					#state{compress=undefined};
+				_ ->
+					#state{compress=gzip}
+			end
+	catch
+		_:_ ->
+			#state{compress=undefined}
+	end.
+
+fold(Commands, State) ->
+	fold(Commands, State, []).
+
+fold([], State, Acc) ->
+	{lists:reverse(Acc), State};
+fold([{response, Status, Headers0, Body}|Tail], State=#state{ignore=false}, Acc) ->
+	Headers = add_accept_encoding(Headers0),
+	fold(Tail, State, [{response, Status, Headers, Body}|Acc]);
+fold([{headers, Status, Headers0} | Tail], State=#state{ignore=false}, Acc) ->
+	Headers = add_accept_encoding(Headers0),
+	fold(Tail, State, [{headers, Status, Headers}|Acc]);
+fold([Command|Tail], State, Acc) ->
+	fold(Tail, State, [Command|Acc]).
+
+add_accept_encoding(Headers=#{<<"accept-encoding">> := AcceptEncoding}) ->
+	try cow_http_hd:parse_accept_encoding(iolist_to_binary(AcceptEncoding)) of
+		List ->
+			case lists:keyfind(<<"gzip">>, 1, List) of
+				%% gzip is excluded but this handler is not ignored; we replace.
+				{_, 0} ->
+					Replaced = lists:keyreplace(<<"gzip">>, 1, List, {<<"gzip">>, 1000}),
+					Codings = build_accept_encoding(Replaced),
+					Headers#{<<"accept-encoding">> => Codings};
+				{_, _} ->
+					Headers;
+				false ->
+					case lists:keyfind(<<"*">>, 1, List) of
+						%% Others are excluded along with gzip; we add.
+						{_, 0} ->
+							WithGzip = [{<<"gzip">>, 1000} | List],
+							Codings = build_accept_encoding(WithGzip),
+							Headers#{<<"accept-encoding">> => Codings};
+						{_, _} ->
+							Headers;
+						false ->
+							Headers#{<<"accept-encoding">> => [AcceptEncoding, <<", gzip">>]}
+					end
+			end
+	catch _:_ ->
+		Headers#{<<"accept-encoding">> => <<"gzip">>}
+	end;
+add_accept_encoding(Headers) ->
+	Headers#{<<"accept-encoding">> => <<"gzip">>}.
+
+%% From cowlib, maybe expose?
+qvalue_to_iodata(0) -> <<"0">>;
+qvalue_to_iodata(Q) when Q < 10 -> [<<"0.00">>, integer_to_binary(Q)];
+qvalue_to_iodata(Q) when Q < 100 -> [<<"0.0">>, integer_to_binary(Q)];
+qvalue_to_iodata(Q) when Q < 1000 -> [<<"0.">>, integer_to_binary(Q)];
+qvalue_to_iodata(1000) -> <<"1">>.
+
+build_accept_encoding([{ContentCoding, Q}|Tail]) ->
+	Weight = iolist_to_binary(qvalue_to_iodata(Q)),
+	Acc = <<ContentCoding/binary, ";q=", Weight/binary>>,
+	do_build_accept_encoding(Tail, Acc).
+
+do_build_accept_encoding([{ContentCoding, Q}|Tail], Acc0) ->
+	Weight = iolist_to_binary(qvalue_to_iodata(Q)),
+	Acc = <<Acc0/binary, ", ", ContentCoding/binary, ";q=", Weight/binary>>,
+	do_build_accept_encoding(Tail, Acc);
+do_build_accept_encoding([], Acc) ->
+	Acc.
+
+inflate(Z, RatioLimit, Data) ->
+	try
+		{Status, Output} = zlib:safeInflate(Z, Data),
+		Size = iolist_size(Output),
+		do_inflate(Z, Size, byte_size(Data) * RatioLimit, Status, [Output])
+	catch
+		error:data_error ->
+			zlib:close(Z),
+			{error, data}
+	end.
+
+do_inflate(Z, Size, Limit, Status, _) when Size > Limit ->
+	case Status of
+		continue -> ok;
+		finished -> zlib:inflateEnd(Z)
```

Comment:
Do you think this is useful, can't we just zlib:close?

Reply:
Yeah we can just close. https://github.com/erlang/otp/blob/master/erts/emulator/nifs/common/zlib_nif.c#L288-L293